### ISSUE TITLE
Move res-client, morph-client, lexicon-client, lemma-client libraries into a parent client-adapters repository

### ISSUE THREAD

**Issue by irina060981** [alpheios-project/components#264]
Refactor adapters to one single adapter
Create documentation for adapters, for adding new adapters
Work on documentation for adding a new language
Try to refactor Lexical Query tool for using it in parallel mode (lexical-test)

**Comment by irina060981** [comment]
**1 step** - Created [client-adapters repo](https://github.com/alpheios-project/client-adapters), uploaded there morph-client and made some refactoring

**Comment by irina060981** [comment]
@kirlat and @balmas , Here it is my suggestion about how to organize client-adapters code:

Client-adapters repo would be attached as a module (similiar to other our modules) with a single import resource - `ClientAdapter`.
And we would be able to use adapters using the following constructions, for example:
``
let homonym = ClientAdapter.maAdapter({
   type: 'getHomonym',
   languageID: CONSTANTS.LANG_LATIN,
   text: 'cepit'
})
``
``
ClientAdapter.lexicons({
  type: 'fetchShortDefs',
  lemma: <some lemma>,
  options: <some lexical options>
})
``

Inside **ClientAdapter** - it would be a simple hierarchy:

`ClientAdapter` - class has static methods for each adapter.
Each adapter has simple class hierarchy:

`BaseAdapter` (common for all adapters within defined fetching methods for windows and node)
`TransformAdapter` (for now it is unique for each one and transforms retrieved data to that one we need)
`ConfigData` (for now it is unique for each adapter and loads different config and constants data)

Current `Adapter` extends from `BaseAdapter` - unique for each adapter's type

I have only started to work on all it and try to define common parts.

My suggestion allows to simply operate using ClientAdapter with simple configuration data, passing into it, it doesn't save any additional data to memory and I think it easily allows to add different other adapters.

**Comment by kirlat** [comment]
@irina060981, what would `ClientAdapter.<Adapter>()` (e.g. `ClientAdapter.maAdapter()`) exactly do behind the scenes? Would it create an instance of a corresponding adapter or reuse an existing instance if such instance is already created, run a query and return a result? I think in such case this function should be asynchronous as that's the usual nature of such calls. What do you think?

**Comment by kirlat** [comment]
If we would be able to implement calls to all adapters using a unified API, that would bring many benefits to us, I think. Actually, this is a great idea! We should just be careful to maintain clear separation between adapter-specific pieces of logic within an adapter library. Adapter-specific code should not interweave because that will be a mess.

The only thing is that if we need to make several request to the same adapter this can be tedious: 
```
let homonym1 = ClientAdapter.maAdapter({ type: 'getHomonym', languageID: CONSTANTS.LANG_LATIN, text: 'text1' })
let homonym2 = ClientAdapter.maAdapter({ type: 'getHomonym', languageID: CONSTANTS.LANG_LATIN, text: 'text2' })
let homonym3 = ClientAdapter.maAdapter({ type: 'getHomonym', languageID: CONSTANTS.LANG_LATIN, text: 'text3' })
```
One of the solutions to this might be to create some "adapter object" (not necessarily an adapter instance, but probably some kind of proxy that will capture repeating part of parameters to avoid specifying them on every call). It could look like (names are just to show purpose, we could name it any name we want)
```
let adapterObject = ClientAdapter.maAdapterObject({ type: 'getHomonym', languageID: CONSTANTS.LANG_LATIN })
let homonym1 = adapterObject({ text: 'text1' })
let homonym2 = adapterObject({ text: 'text2' })
let homonym3 = adapterObject({ text: 'text3' })
```
adapterObject can be a simple function to capture repeating parameters (this is a pseudo-code):
```
ClientAdapter.maAdapterObject(repeatingParams) {
    return function (params) {
        let combinedParams = object.Assign(repeatingParams, params) // Combine repeating params and params somehow
        retrun ClientAdapter.maAdapter(combinedParams)
    }
}
```
Would that be convenient? What do you think?

There could be many advantages of the approach you offered. For example, we could introduce a discovery service, that might return us available adapter for a given language. Or we could combine results from calls to different adapters if results from one adapter is incomplete (the other adapter would append its data or replace those pieces that are known to be incorrect). Or if one adpter's source is down, we can use an adapter to a different source as a spare alternative. We could even create "vurtual adapters" for some specific purposes (e.g. some specific complex queries). Overall, that could make a really flexible and easy to use architecture!

**Comment by balmas** [comment]
I think this is all going in the right direction :-) I think eliminating redundant code and having a common way to interact with the back-end resources is the right way to go.  

A big thing that is missing from our architecture right now is support for discovery of available resources.   This has always been a goal, but it was postponed in the interest of getting the code out.  If we can standardize this without losing the granularity of the different needs I am all for it.

Right now, the individual client adapter libraries each supply their own sort of "discovery" service to map resources to the request.  In the case of the morph-client, this is pretty simple, mapping language to engine right in the main adapter class, whereas for the lexicon-client, the Lexicons class serves as a manager for individual adapters, instantiating one per language/lexicon combo. 

And the client-adapter libraries don't provide any interface to their clients to enable discovery of the configured resources. And because we need to allow the user to choose between different resources sometimes, this means that there is configuration duplicated in several places. For example, the language-options-default.json for the components repository contains information that is redundant with that of the configuration for the lexicon client. 

A user should be able to define their preferences for lexicons for a specific language, and have that information saved with their data, but the user preferences should be considered in the context of the available resources advertised by the application. (E.g. so that if we change the url for a back-end service, we don't have to worry about the user having saved in their preferences a url from an earlier iteration)

Additionally, different sites or applications might want to define different combinations of back-end services. A specific site might only want to offer a certain lexicon or grammar, for example.

**Comment by irina060981** [comment]
@balmas , what dou you think - could I close this issue as I made all changes from clint-adapters code review?
May be some new functionality (like listing available sources, having hookes to existed adapters) would be better to be tracked upon new detailed issues?

**Comment by balmas** [comment]
yes, I agree! closing.
